### PR TITLE
perf: optimize target_arch switching

### PR DESCRIPTION
**Pull request by borngraced**
#543

### CODE REVIEW

File: lua/rustaceanvim/rust_analyzer.lua [hunk 1]
```diff
@@ -5,18 +5,33 @@ local os = require('rustaceanvim.os')
...
-  return vim.lsp.get_clients(filter)
+  local clients = vim.lsp.get_clients(client_filter)
+  clients = vim.tbl_filter(function(client)
+    local rust_settings = client.config and client.config.settings and client.config.settings['rust-analyzer']
+    local cargo_target = rust_settings and rust_settings.cargo and rust_settings.cargo.target
+    -- If no filter is provided or filter.rustc_target is nil, include the client only if cargo_target exists
+    if not filter or not filter.rustc_target then
+      return cargo_target ~= nil
+    end
+    -- Include the client if cargo_target differs from filter.rustc_target
```
Comment (mrcjkb) [inline lua/rustaceanvim/rust_analyzer.lua:31]:
```suggestion
```

No need for this comment if we rename the filter param to `exclude_rustc_target`.
I prefer longer, more expressive variable names over comments.

File: lua/rustaceanvim/rust_analyzer.lua [hunk 2]
```diff
@@ -5,18 +5,33 @@ local os = require('rustaceanvim.os')
 ---@class rustaceanvim.rust-analyzer.ClientAdapter
 local M = {}
 
+---@class GetClientFilter
```
Comment (mrcjkb) [inline lua/rustaceanvim/rust_analyzer.lua:8]: suggestion: extend the built-in class.

```lua
---@class rustaceanvim.lsp.get_clients.Filter: vim.lsp.get_clients.Filter
---@field rustc_target? string Optional cargo target to filter rust-analyzer clients
```

**Review by mrcjkb — CHANGES_REQUESTED** [review]

File: lua/rustaceanvim/rust_analyzer.lua [hunk 3]
```diff
@@ -5,18 +5,33 @@ local os = require('rustaceanvim.os')
...
   if bufnr then
-    filter.bufnr = bufnr
+    client_filter.bufnr = bufnr
   end
-  return vim.lsp.get_clients(filter)
+  local clients = vim.lsp.get_clients(client_filter)
+  clients = vim.tbl_filter(function(client)
+    local rust_settings = client.config and client.config.settings and client.config.settings['rust-analyzer']
+    local cargo_target = rust_settings and rust_settings.cargo and rust_settings.cargo.target
+    -- If no filter is provided or filter.rustc_target is nil, include the client only if cargo_target exists
```
[the remark this answers is not in the document]
Comment (borngraced) [reply lua/rustaceanvim/rust_analyzer.lua:27]: Yeah, I'm trying to cover all possible scenarios but seems to be too strict and looks like a redundant check. Relaxed

**Comment by borngraced** [comment]
improved target switching and reduced nested callbacks @mrcjkb

File: lua/rustaceanvim/lsp/init.lua [hunk 4]
```diff
@@ -320,44 +288,31 @@ M.reload_settings = function(bufnr)
...
+M.set_target_arch = function(target_arch)
+  -- Load target architectures if not already available
+  if not rust_analyzer.rustc_target_archs or rust_analyzer.os_rustc_target then
+    rust_analyzer.load_target_archs()
+  end
+  target_arch = target_arch or rust_analyzer.os_rustc_target
+  if not rust_analyzer.rustc_target_archs[target_arch] then
+    vim.notify('Invalid target architecture provided: ' .. tostring(target_arch), vim.log.levels.ERROR)
+    return
+  end
```
Comment (mrcjkb) [inline lua/rustaceanvim/lsp/init.lua:302]: issue: This isn't as reliable as the previous async approach backed by a cache (`load_target_archs` loads the targets asynchronously, so it won't help here), and it increases the complexity.

I'm also not a fan of eagerly populating the cache in the `BufEnter` autocommand, as it could lead to errors for people who don't need this feature.

Let's focus this PR on filtering by target when restarting. We can move the cache to the `rust_analyzer` module if needed.

**Review by mrcjkb — CHANGES_REQUESTED** [review]

File: lua/rustaceanvim/lsp/init.lua (hunk 4 above)
Comment (borngraced) [reply lua/rustaceanvim/lsp/init.lua]: @mrcjkb  Thanks for the feedback! I've simplified it to just handle restart filtering now. Sorry for the delay in responding - it's been a busy work week!

File: lua/rustaceanvim/lsp/init.lua [hunk 5]
```diff
@@ -320,32 +323,19 @@ M.reload_settings = function(bufnr)
...
-
+  restart(exclude_rustc_target, function(client)
     with_rustc_target_architectures(function(rustc_targets)
-      if target == nil or rustc_targets[target] then
-        client.settings['rust-analyzer'].cargo.target = target
-        client.notify('workspace/didChangeConfiguration', { settings = client.config.settings })
-        vim.notify('Target architecture updated successfully to: ' .. target, vim.log.levels.INFO)
+      if rustc_targets[exclude_rustc_target] then
+        client.settings['rust-analyzer'].cargo.target = exclude_rustc_target
+        client.notify('workspace/didchangeconfiguration', { settings = client.config.settings })
```
Comment (mrcjkb) [inline lua/rustaceanvim/lsp/init.lua:334]: issue: it appears a lot of code was accidentally lowercased (see also the log levels in this file).

```suggestion
        client.notify('workspace/didChangeConfiguration', { settings = client.config.settings })
```

File: lua/rustaceanvim/lsp/init.lua [hunk 6]
```diff
@@ -354,10 +344,9 @@ end
 
 ---Restart the LSP client.
 ---Fails silently if the buffer's filetype is not one of the filetypes specified in the config.
----@param bufnr? number The buffer number (optional), defaults to the current buffer
```
Comment (mrcjkb) [inline lua/rustaceanvim/lsp/init.lua:357]: question: is there a reason for removing this parameter? It seems out of scope for this PR.

File: lua/rustaceanvim/rust_analyzer.lua [hunk 7]
```diff
@@ -5,18 +5,82 @@ local os = require('rustaceanvim.os')
...
 local M = {}
 
+--- Default target value for rustc when no specific target is provided.
+--- Used as a fallback to let rustc determine the appropriate target based on the OS.
+DEFAULT_RUSTC_TARGET = 'OS'
+
+---Local rustc targets cache
+local rustc_targets_cache = nil
+
+M.load_os_rustc_target = function()
```
Comment (mrcjkb) [inline lua/rustaceanvim/rust_analyzer.lua:15]: issue: this appears to be unused.

File: lua/rustaceanvim/rust_analyzer.lua [hunk 8]
```diff
@@ -5,18 +5,82 @@ local os = require('rustaceanvim.os')
...
+        end
+      end
+    end
+  end)
+end
+
+---Handles retrieving rustc target architectures and running the passed in callback
+---to perform certain actions using the retrieved targets.
+---@param callback fun(targets: string[])
+M.with_rustc_target_architectures = function(callback)
```
Comment (mrcjkb) [inline lua/rustaceanvim/rust_analyzer.lua:32]: issue: this isn't actually used. There's a duplicate in the `lsp` module that is used.
I think the `lsp` module (local function) or a `rustaceanvim.rustc` module is a better place for it, because this function doesn't actually interact with rust-analyzer.

**Review by mrcjkb — CHANGES_REQUESTED** [review]

File: lua/rustaceanvim/lsp/init.lua (hunk 4 above)
Comment (mrcjkb) [reply lua/rustaceanvim/lsp/init.lua]: no worries, I'm in no rush 😄

File: lua/rustaceanvim/rust_analyzer.lua (hunk 8 above)
Comment (borngraced) [reply lua/rustaceanvim/rust_analyzer.lua:32]: agreed!

File: lua/rustaceanvim/lsp/init.lua (hunk 6 above)
Comment (borngraced) [reply lua/rustaceanvim/lsp/init.lua:357]: the param seems obsolete and wasn't actually used or set

File: lua/rustaceanvim/rust_analyzer.lua (hunk 7 above)
Comment (borngraced) [reply lua/rustaceanvim/rust_analyzer.lua:8]: removed

File: lua/rustaceanvim/rust_analyzer.lua [hunk 9]
```diff
@@ -5,18 +5,46 @@ local os = require('rustaceanvim.os')
...
   })
   if bufnr then
-    filter.bufnr = bufnr
+    client_filter.bufnr = bufnr
+  end
+  local clients = vim.lsp.get_clients(client_filter)
+  if filter and filter.exclude_rustc_target then
+    clients = vim.tbl_filter(function(client)
+      local cargo_target = vim.tbl_get(client, 'config', 'settings', 'rust-analyzer', 'cargo', 'target')
+      if filter.exclude_rustc_target == DEFAULT_RUSTC_TARGET and cargo_target == nil then
```
Comment (mrcjkb) [inline lua/rustaceanvim/rust_analyzer.lua:40]: issue: `DEFAULT_RUSTC_TARGET` is in the `rustc` module (ditto for the other usages).

tip: If you have Nix installed, I recommend you enter a nix devShell (`nix develop`) and run `pre-commit run --all-files`.
It will run all the type checks and lints needed to catch these issues.
Otherwise, you can run `luacheck` and the `lua-language-server` CLI to get the lints without using Nix.

File: lua/rustaceanvim/lsp/init.lua [hunk 10]
```diff
@@ -285,10 +255,14 @@ end
...
 ---@param bufnr? number The buffer number, defaults to the current buffer
+---@param exclude_rustc_target? string Cargo target triple (e.g., 'x86_64-unknown-linux-gnu') to filter rust-analyzer clients
 ---@return vim.lsp.Client[] clients A list of clients that will be stopped
-M.stop = function(bufnr)
+M.stop = function(bufnr, exclude_rustc_target)
   bufnr = bufnr or vim.api.nvim_get_current_buf()
-  local clients = rust_analyzer.get_active_rustaceanvim_clients(bufnr)
+  local clients = rust_analyzer.get_active_rustaceanvim_clients(
+    bufnr,
+    { exclude_rustc_target = exclude_rustc_target or DEFAULT_RUSTC_TARGET }
```
Comment (mrcjkb) [inline lua/rustaceanvim/lsp/init.lua:264]: issue: The `stop` function shouldn't set a rustc target by default, otherwise it will always try to filter, even when we don't need it to. 

```suggestion
    { exclude_rustc_target = exclude_rustc_target }
```

File: lua/rustaceanvim/lsp/init.lua [hunk 11]
```diff
@@ -322,30 +296,18 @@ end
 
 ---Updates the target architecture setting for the LSP client associated with the given buffer.
 ---@param bufnr? number The buffer number, defaults to the current buffer
----@param target? string The target architecture. Defaults to nil(the current buffer's target if not provided).
-M.set_target_arch = function(bufnr, target)
+---@param exclude_rustc_target? string Cargo target triple (e.g., 'x86_64-unknown-linux-gnu') to filter rust-analyzer clients
+M.set_target_arch = function(bufnr, exclude_rustc_target)
```
Comment (mrcjkb) [inline lua/rustaceanvim/lsp/init.lua:300]: issue: Here (just not in `stop`/`restart`), the parameter should be called `target`, as it's the target we want to set, not a target we want to exclude.

**Review by mrcjkb — CHANGES_REQUESTED** [review]

File: lua/rustaceanvim/rust_analyzer.lua (hunk 9 above)
Comment (borngraced) [reply lua/rustaceanvim/rust_analyzer.lua:40]: thanks!

File: lua/rustaceanvim/lsp/init.lua (hunk 10 above)
Comment (borngraced) [reply lua/rustaceanvim/lsp/init.lua:264]: yeah make sense. thanks

**Review by mrcjkb — APPROVED** [review]
